core/templates/domain/blog-admin/blog-admin-backend-api.service.ts

```diff
@@ -0,0 +1,141 @@
+// Copyright 2021 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Backend api service for fetching the admin data;
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { HttpClient } from '@angular/common/http';
+import { Injectable } from '@angular/core';
+
+interface UserRolesBackendResponse {
+  [role: string]: string;
+}
+
+interface RoleToActionsBackendResponse {
+  [role: string]: string[];
+}
+
+interface ConfigPropertiesBackendResponse {
+  [property: string]: Object;
+}
+
+interface ConfigPropertyValues {
+  'max_number_of_tags_assigned_to_blog_post': number,
+  'list_of_default_tags_for_blog_post': string[]
+}
+
+export interface BlogAdminPageDataBackendDict {
+  'config_properties': ConfigPropertiesBackendResponse;
+  'role_to_actions': RoleToActionsBackendResponse;
+  'updatable_roles': UserRolesBackendResponse;
+}
+
+export interface BlogAdminPageData {
+  configProperties: ConfigPropertiesBackendResponse;
+  roleToActions: RoleToActionsBackendResponse;
+  updatableRoles: UserRolesBackendResponse;
+}
+
+
+@Injectable({
+  providedIn: 'root'
+})
+
+export class BlogAdminBackendApiService {
+  constructor(private http: HttpClient) {}
+
+
+  async getDataAsync(): Promise<BlogAdminPageData> {
+    return new Promise((resolve, reject) => {
+      this.http.get<BlogAdminPageDataBackendDict>(
+        '/blogadminhandler').toPromise().then(response => {
+        resolve({
+          updatableRoles: response.updatable_roles,
+          roleToActions: response.role_to_actions,
+          configProperties: response.config_properties,
+        });
+      }, errorResponse => {
+        reject(errorResponse.error.error);
+      });
+    });
+  }
+
+  private async _postRequestAsync(
+      handlerUrl: string, payload?: Object, action?: string): Promise<void> {
+    return new Promise((resolve, reject) => {
+      this.http.post<void>(
+        handlerUrl, { action, ...payload }).toPromise()
```

Comment:
Based on the types above, it seems both action and payload are optional (i.e. they can be undefined). If so, what happens to the request body here?

Reply:
Payload should be there in all cases. Fixed. `action` will not be provided when `updateUserRoleAsync` calls the function.

Reply:
Cool, thanks!